### DISCUSSION TITLE
API proposal: List<T>.CapacityGrowthPercentage

### DISCUSSION THREAD

**Discussion by chucker** [dotnet/runtime#53299] [Ideas]
## Background and Motivation

`List<T>` has a `Capacity` property, which core sounds with its underlying array’s length. It defaults to **4**, but can be manually set in a constructor overload or in the aforementioned property.

When adding items, `EnsureCapacity` will, if needed, grow the capacity so additional items fit. This is currently [hardcoded to *double* the previous value](https://github.com/dotnet/runtime/blob/3764f3781e6dd78e1e8533eb76e1b317e55939d8/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L405). I.e., if you add a fifth item, the capacity becomes **8**. Doubling was presumably chosen as an adequate compromise for typical use. Note, however, that capacity increases require copying the entire existing list to a new internal array.

## Proposed API
```
public short CapacityGrowthPercentage { get; set; }
```

## Usage examples
In most cases, you’ll use `List<T>` as before. The *default* `CapacityGrowthPercentage` is **100**; the capacity grows by 100% each time.

For cases where you anticipate a list that
* is large to begin with,
* may grow, but not by huge amounts,
you might set it to a much lower value:

```
var listWithManyItems = new List(new[] { 0, 1, 2, 3 })
{
    CapacityGrowthPercentage = 10
}

// grows the capacity from its default 4
// by 10%, rounded up, to 5
listWithManyItems.Add(4);
```

**Comment by davidfowl** [comment]
Where did you run into this being a problem in practice?

**Comment by lechu445** [comment]
similar issue https://github.com/dotnet/runtime/issues/53128
